feat(cargo): support range=update-lockfile

szlend opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Adds `update-lockfile` rangeStrategy support to the `cargo` manager.

## Context

As requested in https://github.com/renovatebot/renovate/issues/22820

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=23968) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=23968) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=23968) it.</sub>

szlend commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -64,11 +82,25 @@ export async function updateArtifacts({
+
+    if (isLockFileMaintenance) {
+      await cargoUpdate(packageFileName, true, config.constraints?.rust)
+    } else if (updatedDeps.find(dep => !dep.currentVersion || !dep.newVersion)) {
+      await cargoUpdate(packageFileName, false, config.constraints?.rust)
```
Can se assume `currentVersion` and `newVersion` are always there? The only reason I handle this is because exists unit tests don't include them. I guess `currentVersion` could be missing or incorrect if there is no `Cargo.lock` file present.

rarkins replied:
If a lock file is present then the `currentVersion` should have been derived from it (via `lockedVersion`) and then you can assume you find it in the lock file. However it's also good to code defensively anytime it is easy to do!

szlend replied:
Yeah I guess the only cases where it wouldn't be present is if the lock file was missing, or the lock file and package files were desynced. I'm not really sure what the right approach is if we have a mixture of deps where some have `currentVersion` and some don't.

What we could do:
1. Update the entire lock file
2. Update just the deps that have `currentVersion` and ignore the rest
3. Bail, as this most likely means that the package and lock files are desynced

rarkins replied:
If the lock file is missing then you won't get to this "artifacts updating" logic

viceice reviewed (commented):
please sign the cla, otherwise we can't review

szlend commented on lib/modules/manager/cargo/types.ts:
```diff
@@ -38,3 +38,31 @@ export interface CargoRegistries {
+  version: 3;
+}
+
+export type LockFile = LockFile3 | LockFile1 | LockFile2;
```
It would probably be fine to merge all three versions as I don't think they don't differ in any way that matters for us.

szlend commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -29,6 +29,27 @@ async function cargoUpdate(
+  let cmd = `cargo update --manifest-path ${quote(manifestPath)}`;
+  cmd += ` --package ${depName}@${lockedVersion}`;
+
+  if (newVersion) {
+    cmd += ` --precise ${newVersion}`;
+  }
```
I could see this fail:
-  we depend on `foo` with  `2.0.1`
- the latest version from crates.io is detected as `2.0.2`
- another package in `Cargo.lock` depends on `foo` with  `=2.0.1` (exact version)

Renovate will try and update `foo` to `2.0.2` at which point `cargo update` will fail saying there's no way to resolve the dependencies given these constraints.

For this to work, the cargo versioning module would need to be aware of `Cargo.lock` as well.

szlend replied:
As predicted, I started testing this on some real-world repos and ran into this exact issue:
```
Updating crates.io index
error: failed to select a version for `wasm-bindgen`.
    ... required by package `js-sys v0.3.64`
    ... which satisfies dependency `js-sys = "^0.3"` of package `nickel-lang-core v0.1.0`
    ... which satisfies path dependency `nickel-lang-core` (locked to 0.1.0) of package `nickel-lang-cli v1.0.0`
versions that meet the requirements `^0.2.87` are: 0.2.87
all possible versions conflict with previously selected packages.
  previously selected package `wasm-bindgen v0.2.83`
    ... which satisfies dependency `wasm-bindgen = "=0.2.83"` of package `nickel-lang-core v0.1.0`
    ... which satisfies path dependency `nickel-lang-core` (locked to 0.1.0) of package `nickel-lang-cli v1.0.0`
failed to select a version for `wasm-bindgen` which could resolve this conflict
```

Any ideas what we should do in cases where we can't upgrade to `newVersion`? Do we need to adjust the versioning module to take the lock file into account? Is that even possible with the current API?

rarkins replied:
So first of all, it's perfectly OK to raise a PR but with failing artifacts. Renovate makes it very clear using both a failing check run as well as a PR comment. Users have the ability to suppress failing PRs if they really wish, but usually it's better to know.

However it's also OK for you to add an algorithm to catch and retry/refine the update if the output from cargo is clear enough. e.g. try updating only the selected dependency/dependencies first, but then update more of them in another retry. 

I did something similar for bundler here: https://github.com/renovatebot/renovate/blob/76c9c63417d72e38dac7d00040c33186f9df4dad/lib/modules/manager/bundler/artifacts.ts#L248-L272

szlend replied:
I think for the initial version I'd rather just open a PR with failing artifacts. The issue is that this instance this will most likely result in an empty commit, so no PR will actually be created. I added additional logging to `cargoUpdatePrecise` so it's more obvious which package failed to update and why.

viceice commented on lib/modules/manager/cargo/artifacts.spec.ts:
```diff
@@ -113,6 +113,32 @@ describe('modules/manager/cargo/artifacts', () => {
+        config,
+      })
+    ).not.toBeNull();
+    expect(execSnapshots).toMatchSnapshot();
```
no new snapshot please.
https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#unit-testing

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -64,11 +85,27 @@ export async function updateArtifacts({
+      await cargoUpdate(packageFileName, true, config.constraints?.rust);
+    } else if (updatedDeps.find(dep => dep.lockedVersion)) {
+      // If we can identify packages by their locked version, then update them precisely
+      for (const dep of updatedDeps) {
```
move loop inside the function and pass all commands to a single exec call, you can pass a string array (perf)

otherwise it would cause multiple docker calls and multiple rust installs on each exec on docker binary source.

szlend replied:
Unfortunately you can't batch cargo updates if `--precise` is specified. I guess we could batch it ourselves through `sh`/`bash` though.

viceice replied:
our exec already supports batches, just pass an array with all required commands and renovate will do it.

viceice commented on lib/modules/manager/cargo/extract.spec.ts:
```diff
@@ -258,5 +260,14 @@ tokio = { version = "1.21.1" }`;
+      await writeLocalFile('Cargo.lock', cargo7lock);
+
+      const res = await extractPackageFile(cargo7toml, 'Cargo.toml', config);
+      expect(res?.deps).toMatchSnapshot();
```
see other comment

viceice commented on lib/modules/manager/cargo/extract.ts:
```diff
@@ -234,7 +236,23 @@ export async function extractPackageFile(
   const res: PackageFileContent = { deps };
   // istanbul ignore if
```
remove this

viceice commented on lib/modules/manager/cargo/extract.ts:
```diff
@@ -234,7 +236,23 @@ export async function extractPackageFile(
+
+    const versionsByPackage = await extractLockFileVersions(lockFileName);
+    for (const dep of deps) {
+      const versions = versionsByPackage.get(`${dep.depName!}`) ?? [];
```
```suggestion
      const versions = versionsByPackage.get(dep.depName!) ?? [];
```
don't wrap

viceice commented on lib/modules/manager/cargo/extract.ts:
```diff
@@ -234,7 +236,23 @@ export async function extractPackageFile(
+    const versionsByPackage = await extractLockFileVersions(lockFileName);
+    for (const dep of deps) {
+      const versions = versionsByPackage.get(`${dep.depName!}`) ?? [];
+      const lockedVersion = cargoVersioning.getSatisfyingVersion(
```
should have a try catch, this function can throw on invalid versions.

viceice commented on lib/modules/manager/cargo/locked-version.ts:
```diff
@@ -0,0 +1,22 @@
+    const content = await readLocalFile(lockFilePath);
+    const lock = parseLockFile(content!.toString())!;
+    const packages = lock.package ?? [];
+    packages.forEach((pkg) => {
```
use for-of loop

viceice commented on lib/modules/manager/cargo/types.ts:
```diff
@@ -38,3 +38,30 @@ export interface CargoRegistries {
   // maps registry names to URLs
   [key: string]: string;
 }
+
```
use zod schema instead, checkout latest poetry refactoring by @zharinov

szlend replied:
Moved the lockfile to `zod`. Ideally we'd move `Cargo.toml` to it as well, though I'd prefer to focus on `update-lockfile` in this MR.

szlend commented:
Should I also update the default (`auto`) range strategy now that we support `update-lockfile`?

viceice commented:
> Should I also update the default (`auto`) range strategy now that we support `update-lockfile`?

yes and please don't force push. we squash merge, so no worry about git history

viceice commented on lib/modules/manager/cargo/artifacts.spec.ts:
```diff
@@ -113,6 +113,36 @@ describe('modules/manager/cargo/artifacts', () => {
+        newPackageFileContent: '{}',
+        config,
+      })
+    ).not.toBeNull();
```
use `toEqual`, otherwise we miss silent artifact errors

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -29,6 +31,46 @@ async function cargoUpdate(
+      (dep) =>
+        `cargo update --manifest-path ${quote(manifestPath)}` +
+        ` --package ${dep.packageName!}@${dep.lockedVersion}` +
+        (dep.newVersion ? ` --precise ${dep.newVersion}` : '')
```
i don't think newVersion can be missing here. Renovate would filter those before calling manager

viceice reviewed (commented):
needs more tests for coverage

szlend commented on lib/modules/manager/cargo/range.ts:
```diff
@@ -11,5 +11,5 @@ export function getRangeStrategy({
     return 'widen';
   }
-  return 'bump';
+  return 'update-lockfile';
```
Reverted

szlend commented on lib/modules/manager/cargo/extract.ts:
```diff
@@ -286,9 +288,33 @@ export async function extractPackageFile(
+          dep.lockedVersion = lockedVersion;
+        }
+      } catch (err) {
+        // istanbul ignore next: this should not happen
+        logger.warn(
+          { packageName, err },
+          `Error finding satisfying version for package ${dep.depName}.`
+        );
```
Ignoring this one because I can't really tell when this could happen.

szlend commented:
Oops, fixed the linting error from a last minute change. Should pass CI now.

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -29,6 +31,44 @@ async function cargoUpdate(
+      const msg = err.stderr.match(regEx(/error: (.+)/))?.at(1);
+      if (pkg && msg) {
+        logger.warn(
+          { err },
```
```suggestion
```
will be logged on outer catch, so no duplicate

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -29,6 +31,44 @@ async function cargoUpdate(
+      const pkg = err.cmd.match(regEx(/--package ([^@]+)@/))?.at(1);
+      const msg = err.stderr.match(regEx(/error: (.+)/))?.at(1);
+      if (pkg && msg) {
+        logger.warn(
```
I think we should only debug log

szlend replied:
Will debug logs show up in PRs/Dependency dashboard? That's the reason I added this initially.

viceice replied:
that's why we've artifact errors, they will logged as comment to the PR. so you should return an artifact error with the proper message then.

szlend replied:
I think the issue is that no PR will be opened if the commit is empty. Are artifact errors also summed up in the Dashboard?

szlend replied:
@viceice Though if this is a known renovate-wide issue, and `logger.warn` is not supposed to be used for things like this, then I'm fine with making this change.

viceice replied:
return as artifact error, this will give you a comment on the PR.

szlend replied:
@viceice Yes I can do that, but you didn't really address my concern about empty commits.

viceice replied:
artifact errors should be visible on dashboard

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -29,6 +31,44 @@ async function cargoUpdate(
+          `Could not update cargo package \`${pkg}\`: ${msg}`
+        );
+      } else if (pkg) {
+        logger.warn({ err }, `Could not update cargo package \`${pkg}\`.`);
```
```suggestion
        logger.debug(`Could not update cargo package \`${pkg}\`.`);
```

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -64,11 +104,21 @@ export async function updateArtifacts({
+
+    if (isLockFileMaintenance) {
+      await cargoUpdate(packageFileName, true, config.constraints?.rust);
+    } else if (updatedDeps.find((dep) => dep.lockedVersion)) {
```
what if there are only some updates with a locked version? can that happen? they would now be silently ignored

szlend replied:
The only case I can think of is if `Cargo.lock` file is not in sync with `Cargo.toml`. Imo this is fine as it's a user mistake and will be corrected the next time the lock file is touched by cargo.

And I can't really think of a good way to handle this anyway. If I regenerate the lock file before trying to precisely update dependencies, then the `lockedVersion` of other dependencies will no longer match the state in `Cargo.lock` an we'll have no way to reference dependencies for a precise update.

viceice replied:
we should probably return an artifact error If not all updates have a locked version. or we do a full lockfile update as we do currently and only log a warning.

szlend replied:
Decided to do a regular workspace update and log a warning.

szlend commented:
This is going to need more work. Just realized this strategy doesn't work when an update to one dependency causes an update to another and as a result can no longer be referenced by the old locked version.

szlend commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -29,6 +29,28 @@ async function cargoUpdate(
+    .filter((dep) => !!dep.lockedVersion)
+    .map(
+      (dep) =>
+        `cargo update --manifest-path ${quote(manifestPath)}` +
+        ` --package ${dep.packageName!}@${dep.lockedVersion}` +
+        ` --precise ${dep.newVersion}`
```
@viceice How do you feel about doing something like this to handle dependencies that have already been updated to the new version? The alternative would be to add recursive retries to `updateArtifacts`, but I feel like that's way more complicated.

```suggestion
        // Hack: If a package is already at `newVersion` then skip updating.
        `cargo update --manifest-path ${quote(manifestPath)}` +
        ` --package ${dep.packageName!}@${dep.newVersion}` +
        ` --precise ${dep.newVersion}` +
        ' > /dev/null 2>&1 ||' +
        // Otherwise update the package to `newVersion`.
        ` cargo update --manifest-path ${quote(manifestPath)}` +
        ` --package ${dep.packageName!}@${dep.lockedVersion}` +
        ` --precise ${dep.newVersion}`
```

szlend replied:
Got no response so I went ahead and implemented it like this.

baraknaveh commented:
@szlend Ht would be awesome to have this ready! Is there much work remaining to get this done?

szlend commented:
- Added a hack to handle already updated dependencies. This is necessary when a preceding cargo command updates a later dependency.
- Implemented `updateLockedDependency` which improved error reporting significantly.

I also tested this on a number of personal projects and haven't noticed any issues other than the previously mentioned `Detected empty commit` message in the Dashboard when something fails. Please re-review.

rarkins reviewed (commented):
Let's say that a PR already exists and is up-to-date. Will Renovate re-run `cargo update` commands every single time it runs on the repo/branch, or does it have a way of knowing that the locked version is already updated and no command is necessary?

szlend commented:
> Let's say that a PR already exists and is up-to-date. Will Renovate re-run `cargo update` commands every single time it runs on the repo/branch, or does it have a way of knowing that the locked version is already updated and no command is necessary?

Is this something each `manager` has to handle itself? I implemented the `'already-updated'` case in `updateLockedDependency` so technically Renovate should be aware of the state of the lockfile at least.

rarkins commented:
I think you have it covered already with the `already-updated` part. You can verify that by running Renovate twice - first time to create the branch and second to check that no `cargo update` command was run

szlend commented:
> I think you have it covered already with the `already-updated` part. You can verify that by running Renovate twice - first time to create the branch and second to check that no `cargo update` command was run

The issue is within the same run. Between batched cargo update commands.

viceice commented on lib/modules/manager/cargo/extract.spec.ts:
```diff
@@ -513,5 +519,40 @@ replace-with = "mcorbin"
+      await writeLocalFile('Cargo.lock', cargo7lock);
+
+      const res = await extractPackageFile(cargo7toml, 'Cargo.toml', config);
+      expect(res?.deps[0].lockedVersion).toBe('1.0.1');
+      expect(res?.deps[1].lockedVersion).toBe('2.0.1');
+      expect(res?.deps).toHaveLength(2);
```
same

viceice commented on lib/modules/manager/cargo/extract.spec.ts:
```diff
@@ -513,5 +519,40 @@ replace-with = "mcorbin"
+      await writeLocalFile('Cargo.lock', cargo8lock);
+
+      const res = await extractPackageFile(cargo8toml, 'Cargo.toml', config);
+      expect(res?.deps[0].lockedVersion).toBe('2.0.1');
+      expect(res?.deps[1].lockedVersion).toBeUndefined();
+      expect(res?.deps).toHaveLength(2);
```
same

szlend replied:
How do you expect me to match on non-presence of `lockedVersion` using `toMatchObject`?

This is not an improvement imo:
```javascript
expect(res?.deps).toMatchObject([
  { lockedVersion: '2.0.1' },
  expect.not.objectContaining({ lockedVersion: expect.anything() }),
]);
```

rarkins replied:
If you use `toEqual()` then it must exactly match the contents

szlend replied:
Sure, but then it's asserting on fields that aren't really relevant to the test. I still feel the original code was fine.

rarkins replied:
OK. I'm ok with toMatchObject() if you think some fields are irrelevant

viceice commented on lib/modules/manager/cargo/extract.spec.ts:
```diff
@@ -513,5 +519,40 @@ replace-with = "mcorbin"
+      await writeLocalFile('Cargo.lock', 'foo');
+
+      const res = await extractPackageFile(cargo8toml, 'Cargo.toml', config);
+      expect(res?.deps[0].lockedVersion).toBeUndefined();
+      expect(res?.deps[1].lockedVersion).toBeUndefined();
+      expect(res?.deps).toHaveLength(2);
```
same

viceice commented on lib/modules/manager/cargo/locked-version.ts:
```diff
@@ -0,0 +1,40 @@
+  const versionsByPackage = new Map<string, string[]>();
+  const content = await readLocalFile(lockFilePath, 'utf8');
+  if (!content) {
+    return versionsByPackage;
```
return null on error and log debug message on calling code 
```suggestion
    return null;
```

viceice commented on lib/modules/manager/cargo/locked-version.ts:
```diff
@@ -0,0 +1,40 @@
+    return versionsByPackage;
+  }
+  const packages = lock.package ?? [];
+  for (const pkg of packages) {
```
```suggestion
  for (const pkg of coerceArray(lock.package)) {
```

viceice commented on lib/modules/manager/cargo/update-locked.ts:
```diff
@@ -0,0 +1,23 @@
+    `cargo.updateLockedDependency: ${depName}@${currentVersion} -> ${newVersion} [${lockFile}]`
+  );
+  try {
+    const locked = extractLockFileContentVersions(lockFileContent ?? '');
```
```suggestion
    const locked = extractLockFileContentVersions(lockFileContent!);
```
should never null or undefined here.

szlend commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -33,6 +33,45 @@ async function cargoUpdate(
+          ` --manifest-path ${quote(manifestPath)}` +
+          ` --package ${dep.packageName!}@${dep.newVersion}` +
+          ` --precise ${dep.newVersion}` +
+          ' > /dev/null 2>&1 || ' +
```
I assume the `||` operator doesn't work either then? I don't have a Windows machine to test.

szlend replied:
Solved using recursive updateArtifacts strategy.

baraknaveh commented:
This would be such a great improvement for Cargo deps -- what is left to be solved?

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -33,24 +35,47 @@ async function cargoUpdate(
-  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+async function cargoUpdatePrecise(
+  manifestPath: string,
+  updatedDeps: Upgrade<Record<string, unknown>>[],
```
```suggestion
  updatedDeps: Upgrade[],
```

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -65,14 +90,49 @@ export async function updateArtifacts({
+  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+  if (
+    !isLockFileMaintenance &&
+    (updatedDeps === undefined || updatedDeps.length < 1)
```
```suggestion
    (!updatedDeps?.length)
```
isn't that the same but shorter?

szlend replied:
I guess. It's not really obvious that this is checking for `length != 0` unless you're intimate with javascript.

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -93,7 +153,43 @@ export async function updateArtifacts({
+    if (
+      recursionLimit > 0 &&
+      newCargoLockContent &&
+      String(err.stderr).match(/error: package ID specification/)
```
```suggestion
      regEx(/error: package ID specification/).test(err.stderr)
```

viceice commented on lib/modules/manager/cargo/artifacts.ts:
```diff
@@ -33,24 +36,47 @@ async function cargoUpdate(
+  await exec(cmds, execOptions);
+}
+
+export async function updateArtifacts(
+  {
+    packageFileName,
+    updatedDeps,
+    newPackageFileContent,
+    config,
+  }: UpdateArtifact,
+  recursionLimit = 10
+): Promise<UpdateArtifactsResult[] | null> {
```
please rename and make private, create a new `updateArtifacts` which called the renamed function, so recursion limit isn't part of the public API.

szlend replied:
This is exactly how it's done in other managers, i.e. bundler. I'm really getting tired of this back and forth. Please be respectful of my time. I will leave some time for you to do a final review pass. Any other issues you can fix yourself. I'm happy to use this branch for my use case.

rarkins replied:
I think it's best that we close this and someone else can pick it up to completion

LunNova commented:
This is an important feature for rust developers. Since it's nearly done and has some issues that seem like nitpicks from the whitesource team would it not be easier for one of you to pick up this PR and get it finalized and merged?

Risking dropping the hard work @szlend put in would be really disappointing. I don't personally have the time to sink into working on this myself but I've seen this sort of thing play out a few times in open source projects and it can be really damaging to people's desires to contribute. :(

baraknaveh commented:
I second @LunNova! If the feature is done and there are only some nitpicky follow ups, how about merging it and cut a follow up for the nitpicks?

rarkins commented:
"It's just a trivial amount of work, so someone else should do it".

If you have nothing to contribute, please don't create noise in this repo. If something is important to you then demonstrate with a PR

baraknaveh commented:
@viceice very disappointing. Uninviting to contributors.
